From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: John Galt <johngaltfirstrun@gmail.com>
Date: Wed, 25 Feb 2026 14:42:10 -0500
Subject: [PATCH 04/13] Launcher3: Optimize AllApps perf and icon allocations

Only lookup adapter once per frame, enable LAYER_TYPE_HARDWARE, do not
create new bitmap if already the same
---
 src/com/android/launcher3/BubbleTextView.java | 12 ++++++++-
 .../allapps/AllAppsTransitionController.java  |  5 ++--
 .../RecyclerViewAnimationController.java      | 25 ++++++++++++++++---
 .../android/launcher3/icons/IconCache.java    |  9 ++++++-
 4 files changed, 44 insertions(+), 7 deletions(-)

diff --git a/src/com/android/launcher3/BubbleTextView.java b/src/com/android/launcher3/BubbleTextView.java
index 553db6fd7..cfe605aea 100644
--- a/src/com/android/launcher3/BubbleTextView.java
+++ b/src/com/android/launcher3/BubbleTextView.java
@@ -1357,6 +1357,11 @@ public class BubbleTextView extends TextView implements ItemInfoUpdateReceiver,
         return mIsIconVisible ? mIcon : new ColorDrawable(Color.TRANSPARENT);
     }
 
+    @Override
+    public boolean hasOverlappingRendering() {
+        return false;
+    }
+
     /** Sets the icon visual state to disabled or not. */
     public void setIconDisabled(boolean isDisabled) {
         if (mIcon != null) {
@@ -1374,7 +1379,9 @@ public class BubbleTextView extends TextView implements ItemInfoUpdateReceiver,
         // same as before.
         mDisableRelayout = mIcon != null;
 
-        icon.setBounds(0, 0, mIconSize, mIconSize);
+        if (icon.getBounds().width() != mIconSize || icon.getBounds().height() != mIconSize) {
+            icon.setBounds(0, 0, mIconSize, mIconSize);
+        }
 
         updateIcon(icon);
 
@@ -1428,6 +1435,9 @@ public class BubbleTextView extends TextView implements ItemInfoUpdateReceiver,
         CacheLookupFlag expectedFlag = DEFAULT_LOOKUP_FLAG.withThemeIcon(shouldUseTheme());
         if (getTag() instanceof ItemInfoWithIcon info && !mHighResUpdateInProgress
                 && info.getMatchingLookupFlag().isVisuallyLessThan(expectedFlag)) {
+                if (mIcon != null && mIcon.isThemed() && shouldUseTheme()) {
+                    return;
+                }
             if (mIconLoadRequest != null) {
                 mIconLoadRequest.cancel();
             }
diff --git a/src/com/android/launcher3/allapps/AllAppsTransitionController.java b/src/com/android/launcher3/allapps/AllAppsTransitionController.java
index c51805024..89997a26c 100644
--- a/src/com/android/launcher3/allapps/AllAppsTransitionController.java
+++ b/src/com/android/launcher3/allapps/AllAppsTransitionController.java
@@ -313,8 +313,9 @@ public class AllAppsTransitionController
         if (hasScaleEffect != mHasScaleEffect) {
             mHasScaleEffect = hasScaleEffect;
             if (mHasScaleEffect) {
-                modifyAttributesOnViewTree(rv, mLauncher.getAppsView(),
-                        CLIP_CHILDREN_FALSE_MODIFIER);
+                // Do not disable clipping if drawer opening, otherwise we tank frame rate
+                if (scaleProgress < 0.95f)
+                    modifyAttributesOnViewTree(rv, mLauncher.getAppsView(), CLIP_CHILDREN_FALSE_MODIFIER);
             } else {
                 restoreAttributesOnViewTree(rv, mLauncher.getAppsView(),
                         CLIP_CHILDREN_FALSE_MODIFIER);
diff --git a/src/com/android/launcher3/allapps/RecyclerViewAnimationController.java b/src/com/android/launcher3/allapps/RecyclerViewAnimationController.java
index 8752e4a1a..50b7fb9c1 100644
--- a/src/com/android/launcher3/allapps/RecyclerViewAnimationController.java
+++ b/src/com/android/launcher3/allapps/RecyclerViewAnimationController.java
@@ -90,6 +90,8 @@ public class RecyclerViewAnimationController {
         Integer top = null;
         AllAppsRecyclerView allAppsRecyclerView = getRecyclerView();
 
+        List<BaseAllAppsAdapter.AdapterItem> allAppsAdapters = allAppsRecyclerView.getApps().getAdapterItems();
+
         for (int i = 0; i < allAppsRecyclerView.getChildCount(); i++) {
             View currentView = allAppsRecyclerView.getChildAt(i);
             if (currentView == null) {
@@ -99,8 +101,6 @@ public class RecyclerViewAnimationController {
                 top = currentView.getTop();
             }
             int adapterPosition = allAppsRecyclerView.getChildAdapterPosition(currentView);
-            List<BaseAllAppsAdapter.AdapterItem> allAppsAdapters = allAppsRecyclerView.getApps()
-                    .getAdapterItems();
             if (adapterPosition < 0 || adapterPosition >= allAppsAdapters.size()) {
                 continue;
             }
@@ -188,7 +188,22 @@ public class RecyclerViewAnimationController {
             duration = 0;
         }
 
-        mAnimator.addListener(forEndCallback(() -> mAnimator = null));
+        mAnimator.addListener(forEndCallback(() -> {
+            mAnimator = null;
+            AllAppsRecyclerView rv = getRecyclerView();
+            for (int i = 0; i < rv.getChildCount(); i++) {
+                rv.getChildAt(i).setLayerType(View.LAYER_TYPE_NONE, null);
+            }
+        }));
+        mAnimator.addListener(new android.animation.AnimatorListenerAdapter() {
+            @Override
+            public void onAnimationStart(android.animation.Animator animation) {
+                AllAppsRecyclerView rv = getRecyclerView();
+                for (int i = 0; i < rv.getChildCount(); i++) {
+                    rv.getChildAt(i).setLayerType(View.LAYER_TYPE_HARDWARE, null);
+                }
+            }
+        });
         mAnimator.setDuration(duration).setInterpolator(timeInterpolator);
         mAnimator.addListener(forSuccessCallback(onEndRunnable));
         mAnimator.start();
@@ -199,6 +214,10 @@ public class RecyclerViewAnimationController {
     private void onChildAttached(View child) {
         // Avoid allocating hardware layers for alpha changes.
         child.forceHasOverlappingRendering(false);
+        // Use HW layer for newly attached children
+        if (isRunning()) {
+            child.setLayerType(View.LAYER_TYPE_HARDWARE, null);
+        }
         child.setPivotY(0);
         if (getAnimationProgress() > 0 && getAnimationProgress() < 1) {
             // Before the child is rendered, apply the animation including it to avoid flicker.
diff --git a/src/com/android/launcher3/icons/IconCache.java b/src/com/android/launcher3/icons/IconCache.java
index 1f8649be0..c3c87ce75 100644
--- a/src/com/android/launcher3/icons/IconCache.java
+++ b/src/com/android/launcher3/icons/IconCache.java
@@ -72,6 +72,7 @@ import com.android.launcher3.util.ApplicationInfoWrapper;
 import com.android.launcher3.util.CancellableTask;
 import com.android.launcher3.util.ComponentKey;
 import com.android.launcher3.util.DaggerSingletonTracker;
+import com.android.launcher3.util.FlagOp;
 import com.android.launcher3.util.InstantAppResolver;
 import com.android.launcher3.util.PackageUserKey;
 import com.android.launcher3.widget.WidgetSections;
@@ -601,7 +602,13 @@ public class IconCache extends BaseIconCache {
         if (bitmap == null) {
             return getDefaultIcon(user);
         }
-        return bitmap.withFlags(getUserFlagOpLocked(user));
+        FlagOp userFlagOp = getUserFlagOpLocked(user);
+        int currentFlags = bitmap.getFlags();
+        // Skip creating a new BitMap since identical
+        if (currentFlags == userFlagOp.apply(currentFlags)) {
+            return bitmap;
+        }
+        return bitmap.withFlags(userFlagOp);
     }
 
     protected void applyCacheEntry(@NonNull final CacheEntry entry,
